# PCEP Exam Common Errors: A Gotcha Guide

Welcome to this interactive guide designed to highlight common mistakes and tricky concepts that often trip up students taking the PCEP (Certified Entry-Level Python Programmer) exam. For each problem, try to answer the question first, then run the code, and finally reveal the hint, solution, and explanation to deepen your understanding.

Let's dive in!

---

## Data Types, Variables, Basic Input-Output Operations, and Basic Operators

This section focuses on common errors related to Python's core building blocks.

### Problem 1.1

**Question:** What will be the output of the following code? Predict the output before running.

In [1]:
print(5 / 2)
print(5 // 2)

2.5
2


<details>
<summary>Hint</summary>
Python has two division operators. One always returns a float, and the other performs floor division.
</details>

<details>
<summary>Solution</summary>

```
2.5
2
```

</details>

<details>
<summary>Explanation</summary>
The `/` operator performs classic division and always returns a float, even if the result is a whole number (e.g., `4 / 2` would be `2.0`).
The `//` operator performs floor division, which means it divides and then rounds the result down to the nearest whole number (integer). For positive numbers, this effectively truncates the decimal part.
</details>

---

### Problem 1.2

**Question:** What will be the output of the following code if the user enters `10` when prompted?

```python
age = input("Enter your age: ")
print(age + 5)
```

Predict the output before running.

In [2]:
age = input("Enter your age: ")
print(age + 5)


Enter your age: 5


TypeError: can only concatenate str (not "int") to str

<details>
<summary>Hint</summary>
The `input()` function always returns a string, regardless of what the user types.
</details>

<details>
<summary>Solution</summary>

```
TypeError: can only concatenate str (not "int") to str
```

</details>

<details>
<summary>Explanation</summary>
The `input()` function reads all user input as a string. When you try to add an integer (`5`) to a string (`age`), Python raises a `TypeError` because you cannot concatenate a string and an integer directly. To perform arithmetic operations, you must explicitly convert the input string to a numeric type (e.g., `int()` or `float()`).

Corrected code:
```python
age = int(input("Enter your age: "))
print(age + 5)
```
</details>

---

### Problem 1.3

**Question:** What will be the output of the following code? Predict the output before running.

In [3]:
print(3 + 2 * 4)

11


<details>
<summary>Hint</summary>
Recall the order of operations (PEMDAS/BODMAS): Parentheses, Exponents, Multiplication/Division, Addition/Subtraction.
</details>

<details>
<summary>Solution</summary>

```
11
```

</details>

<details>
<summary>Explanation</summary>
In Python, just like in mathematics, multiplication (`*`) has a higher precedence than addition (`+`). Therefore, `2 * 4` is evaluated first, resulting in `8`. Then, `3` is added to `8`, yielding `11`. Many might mistakenly calculate `(3 + 2) * 4 = 20` if they ignore precedence.
</details>

---

### Problem 1.4

**Question:** What will be the value of `a` after the following code executes? Predict the output before running.

In [4]:
a = 10
b = 2
a /= 2 * b
print(a)

2.5


<details>
<summary>Hint</summary>
The right-hand side of an augmented assignment (`op=`) is evaluated completely before the assignment operation takes place.
</details>

<details>
<summary>Solution</summary>

```
2.5
```

</details>

<details>
<summary>Explanation</summary>
The expression `a /= 2 * b` is equivalent to `a = a / (2 * b)`. First, `2 * b` is evaluated: `2 * 2 = 4`. Then, `a` is divided by this result: `10 / 4 = 2.5`. Finally, `a` is updated to `2.5`. This is distinct from `a = a / 2 * b`, which would be `(10 / 2) * 2 = 10.0`.
</details>

---

### Problem 1.5

**Question:** What will be the output of the following code? Predict the output before running.

In [5]:
myVariable = 5
myvariable = 10
print(myVariable)

5


<details>
<summary>Hint</summary>
Python is a case-sensitive language.
</details>

<details>
<summary>Solution</summary>

```
5
```

</details>

<details>
<summary>Explanation</summary>
Python treats `myVariable` and `myvariable` as two distinct variables because of the difference in capitalization. The first line initializes `myVariable` to 5. The second line initializes a *new* variable `myvariable` to 10. When `print(myVariable)` is called, it refers to the first variable, which still holds the value 5.
</details>

---

### Problem 1.6

**Question:** What will be the output of the following code? Predict the output before running.

In [6]:
print(2 == 2.)

True


<details>
<summary>Hint</summary>
Consider Python's type coercion rules during comparisons.
</details>

<details>
<summary>Solution</summary>

```
True
```

</details>

<details>
<summary>Explanation</summary>
Python performs implicit type conversion when comparing different numeric types. `2` is an integer, and `2.` is a float. When compared, Python treats `2` as `2.0` for the comparison, and since `2.0` is indeed equal to `2.0`, the expression evaluates to `True`.
</details>

---

### Problem 1.7

**Question:** What will be the output of the following code? Predict the output before running.

In [7]:
x = "10"
y = "5"
print(x + y)

105


<details>
<summary>Hint</summary>
The data types of `x` and `y` are strings, not numbers. The `+` operator behaves differently for strings than for numbers.
</details>

<details>
<summary>Solution</summary>

```
105
```

</details>

<details>
<summary>Explanation</summary>
When the `+` operator is used with strings, it performs string concatenation, not arithmetic addition. Even though the strings "10" and "5" contain digits, Python treats them as text. Therefore, "10" + "5" results in "105". To perform arithmetic addition, you would need to convert them to numbers first, e.g., `print(int(x) + int(y))`.
</details>

---

### Problem 1.8

**Question:** What will be the output of the following code? Predict the output before running.

In [8]:
print("Hello" " World")

Hello World


<details>
<summary>Hint</summary>
Python has a special rule for adjacent string literals.
</details>

<details>
<summary>Solution</summary>

```
Hello World
```

</details>

<details>
<summary>Explanation</summary>
Python automatically concatenates two or more string literals that are placed next to each other, separated only by whitespace. This is a syntactic sugar often used for breaking long strings into multiple lines or for simple string joining without explicit `+` operators.
</details>

---

### Problem 1.9

**Question:** What will be the output of the following code? Predict the output before running.

In [9]:
print(-7 % 3)
print(7 % -3)
print(-7 % -3)

2
-2
-1


<details>
<summary>Hint</summary>
In Python, the sign of the result of the modulo operation (`%`) is the same as the sign of the *divisor*.
</details>

<details>
<summary>Solution</summary>

```
2
-2
-1
```

</details>

<details>
<summary>Explanation</summary>
Python's modulo operator `%` works such that the result always has the same sign as the divisor (the second operand).

* `-7 % 3`: The divisor is `3` (positive), so the result is positive. `-7 = 3 * (-3) + 2`. The remainder is `2`.
* `7 % -3`: The divisor is `-3` (negative), so the result is negative. `7 = -3 * (-2) + 1`. The remainder is `1`. To match the divisor's sign, `7 = -3 * (-3) - 2`. The remainder is `-2`.
* `-7 % -3`: The divisor is `-3` (negative), so the result is negative. `-7 = -3 * 3 + 2`. The remainder is `2`. To match the divisor's sign, `-7 = -3 * 2 - 1`. The remainder is `-1`.

This behavior can be counter-intuitive compared to some other programming languages where the sign of the result matches the dividend.
</details>

---

## Boolean Values, Conditional Execution, Loops, Lists and List Processing, Logical and Bitwise Operations

This module covers control flow, logical operations, and list manipulation.

### Problem 2.1

**Question:** What will be the output of the following code? Run the code to see the error.

In [10]:
x = 10
if x > 5:
print("x is greater than 5")

IndentationError: expected an indented block after 'if' statement on line 2 (2587771967.py, line 3)

<details>
<summary>Hint</summary>
Python uses indentation to define code blocks. Incorrect indentation will lead to a specific type of error.
</details>

<details>
<summary>Solution</summary>

```
IndentationError: expected an indented block
```

</details>

<details>
<summary>Explanation</summary>
Python relies heavily on consistent indentation to define the scope of code blocks (e.g., inside `if` statements, `for` loops, or function definitions). The `print` statement here is not indented, but it is expected to be part of the `if` block, leading to an `IndentationError`. Correcting it would involve adding 4 spaces (or a tab) before `print("x is greater than 5")`.
</details>

---

### Problem 2.2

**Question:** What will be the output of the following code? Predict the output before running.

In [11]:
print(True or False and False)

True


<details>
<summary>Hint</summary>
The order of operations for logical operators is `not`, then `and`, then `or`.
</details>

<details>
<summary>Solution</summary>

```
True
```

</details>

<details>
<summary>Explanation</summary>
In Python, logical operators have a specific order of precedence: `not` has the highest, followed by `and`, and then `or` has the lowest. So, the expression `True or False and False` is evaluated as `True or (False and False)`.
First, `False and False` evaluates to `False`.
Then, `True or False` evaluates to `True`.
</details>

---

### Problem 2.3

**Question:** What will be the output of the following code if `x` is `1`? Predict the output before running.

In [12]:
x = 1
if x == 1:
    print("One")
if x == 2:
    print("Two")
if x < 5:
    print("Less than five")

One
Less than five


<details>
<summary>Hint</summary>
Consider how `if` statements are evaluated independently versus how `if-elif-else` chains work.
</details>

<details>
<summary>Solution</summary>

```
One
Less than five
```

</details>

<details>
<summary>Explanation</summary>
When using multiple independent `if` statements, each condition is checked regardless of whether a previous `if` condition was true. In this case, `x == 1` is true, so "One" is printed. Then, `x == 2` is false. Finally, `x < 5` is also true (since `1 < 5`), so "Less than five" is printed.
If you intend to have mutually exclusive conditions, you should use `elif` (else if) to ensure that only one block of code is executed:

```python
x = 1
if x == 1:
    print("One")
elif x == 2:
    print("Two")
elif x < 5:
    print("Less than five")
# Output: One (only the first true condition's block executes)
```
</details>

---

### Problem 2.4

**Question:** What will be the output of the following code? Predict the output before running.

In [13]:
for i in range(5):
    print(i, end=" ")

0 1 2 3 4 

<details>
<summary>Hint</summary>
The `range(stop)` function generates numbers up to, *but not including*, the `stop` value.
</details>

<details>
<summary>Solution</summary>

```
0 1 2 3 4
```

</details>

<details>
<summary>Explanation</summary>
The `range(stop)` function generates a sequence of numbers starting from `0` by default, and stopping *before* the `stop` value. So, `range(5)` produces the sequence `0, 1, 2, 3, 4`. If you needed to include `5`, you would use `range(6)`.
</details>

---

### Problem 2.5

**Question:** What will be the output of the following code? Predict the output before running.

In [14]:
my_list = ['a', 'b', 'c']
print(my_list[1])
print(my_list[3])

b


IndexError: list index out of range

<details>
<summary>Hint</summary>
Python list indexes start at 0. The last valid index is `length - 1`.
</details>

<details>
<summary>Solution</summary>

```
b
IndexError: list index out of range
```

</details>

<details>
<summary>Explanation</summary>
In Python, sequence types like lists, strings, and tuples are zero-indexed. This means the first element is at index `0`, the second at `1`, and so on. For `my_list = ['a', 'b', 'c']`:
* `my_list[0]` is `'a'`
* `my_list[1]` is `'b'`
* `my_list[2]` is `'c'`

Attempting to access `my_list[3]` results in an `IndexError` because there is no element at that index (the valid indices are 0, 1, and 2).
</details>

---

### Problem 2.6

**Question:** What will be the output of the following code? Pay close attention to the loop condition and increment. Predict the output before running.

In [15]:
i = 0
while i <= 3:
    print(i, end=" ")
    i += 1

0 1 2 3 

<details>
<summary>Hint</summary>
Consider the first and last values of `i` for which the condition `i <= 3` is true.
</details>

<details>
<summary>Solution</summary>

```
0 1 2 3
```

</details>

<details>
<summary>Explanation</summary>
An 'off-by-one' error occurs when a loop iterates one too many or one too few times. In this `while` loop, the condition `i <= 3` means the loop continues as long as `i` is less than or equal to 3.
* `i = 0`: `0 <= 3` is True, prints 0.
* `i = 1`: `1 <= 3` is True, prints 1.
* `i = 2`: `2 <= 3` is True, prints 2.
* `i = 3`: `3 <= 3` is True, prints 3.
* `i = 4`: `4 <= 3` is False, loop terminates.

The loop correctly prints 0, 1, 2, and 3. A common error would be to use `i < 3` if the intent was to print only 0, 1, 2, or to forget `i += 1` leading to an infinite loop.
</details>

---

### Problem 2.7

**Question:** What will be the output of the following code? Predict the output before running.

In [16]:
for i in range(5):
    if i == 2:
        continue
    if i == 4:
        break
    print(i, end=" ")

0 1 3 

<details>
<summary>Hint</summary>
`continue` skips the rest of the current loop iteration. `break` exits the loop entirely.
</details>

<details>
<summary>Solution</summary>

```
0 1 3
```

</details>

<details>
<summary>Explanation</summary>
* `i = 0`: `i == 2` is False, `i == 4` is False. Prints `0`.
* `i = 1`: `i == 2` is False, `i == 4` is False. Prints `1`.
* `i = 2`: `i == 2` is True. `continue` is executed, skipping the `print(i)` statement for this iteration and moving to the next value of `i`.
* `i = 3`: `i == 2` is False, `i == 4` is False. Prints `3`.
* `i = 4`: `i == 2` is False. `i == 4` is True. `break` is executed, immediately terminating the loop.

Understanding when to use `break` (to exit the loop) versus `continue` (to skip the current iteration) is crucial for correct loop control.
</details>

---

### Problem 2.8

**Question:** What will be the output of the following code? Predict the output before running.

In [17]:
print(5 & 3)
print(5 | 3)
print(5 ^ 3)
print(~5)
print(5 << 1)
print(5 >> 1)

1
7
6
-6
10
2


<details>
<summary>Hint</summary>
Convert the numbers to their binary representations and perform the operations bit by bit. For `~` (bitwise NOT), remember two's complement representation for negative numbers.
</details>

<details>
<summary>Solution</summary>

```
1
7
6
-6
10
2
```

</details>

<details>
<summary>Explanation</summary>
Let's look at the binary representations (using 8 bits for illustration):
`5` in binary: `0000 0101`
`3` in binary: `0000 0011`

* `5 & 3` (Bitwise AND):
    `0000 0101` (`5`)
    `& 0000 0011` (`3`)
    `----------`
    `0000 0001` (`1`)
    Result is `1`.

* `5 | 3` (Bitwise OR):
    `0000 0101` (`5`)
    `| 0000 0011` (`3`)
    `----------`
    `0000 0111` (`7`)
    Result is `7`.

* `5 ^ 3` (Bitwise XOR):
    `0000 0101` (`5`)
    `^ 0000 0011` (`3`)
    `----------`
    `0000 0110` (`6`)
    Result is `6`.

* `~5` (Bitwise NOT):
    This inverts all bits. For positive `x`, `~x` is equivalent to `-(x + 1)`.
    `0000 0101` (`5`)
    Inverting bits: `1111 1010`. In two's complement, this represents `-6`.
    Result is `-6`.

* `5 << 1` (Left Shift):
    Shifts bits to the left by 1 position, equivalent to multiplying by $2^1$.
    `0000 0101` (`5`) shifted left by 1 becomes `0000 1010` (`10`).
    Result is `10`.

* `5 >> 1` (Right Shift):
    Shifts bits to the right by 1 position, equivalent to integer division by $2^1$.
    `0000 0101` (`5`) shifted right by 1 becomes `0000 0010` (`2`).
    Result is `2`.
</details>

---

### Problem 2.9

**Question:** What will be the output of the following code? Predict the output before running.

In [18]:
list1 = [1, 2, 3]
list2 = list1
list2.append(4)
print(list1)
print(list2)

[1, 2, 3, 4]
[1, 2, 3, 4]


<details>
<summary>Hint</summary>
When you assign one list to another using `=`, you are creating an alias, not a copy. Both variables point to the same object in memory.
</details>

<details>
<summary>Solution</summary>

```
[1, 2, 3, 4]
[1, 2, 3, 4]
```

</details>

<details>
<summary>Explanation</summary>
In Python, lists are mutable objects. When you write `list2 = list1`, you are not creating a new list. Instead, `list2` becomes a new reference (or "alias") to the *same* list object that `list1` refers to. Therefore, any changes made through `list2` (like `list2.append(4)`) will also be reflected when accessing the list through `list1`, because they are both looking at the exact same data. To create a shallow copy, you would use `list2 = list1.copy()` or `list2 = list1[:]`.
</details>

---

### Problem 2.10

**Question:** What will be the value of `squares` after the following code executes? Predict the output before running.

In [19]:
numbers = [1, 2, 3, 4]
squares = [x*x for x in numbers if x % 2 == 0]
print(squares)

[4, 16]


<details>
<summary>Hint</summary>
List comprehensions provide a concise way to create lists. Remember the order: `[expression for item in iterable if condition]`.
</details>

<details>
<summary>Solution</summary>

```
[4, 16]
```

</details>
<details>
<summary>Explanation</summary>
The list comprehension `[x*x for x in numbers if x % 2 == 0]` works as follows:
1.  It iterates through each `x` in the `numbers` list.
2.  For each `x`, it checks if the condition `x % 2 == 0` (is `x` even?) is `True`.
3.  If the condition is `True`, the `expression` (`x*x`) is evaluated, and its result is added to the new `squares` list.

* For `x = 1`: `1 % 2 == 0` is `False`. (Skipped)
* For `x = 2`: `2 % 2 == 0` is `True`. `2*2 = 4` is added.
* For `x = 3`: `3 % 2 == 0` is `False`. (Skipped)
* For `x = 4`: `4 % 2 == 0` is `True`. `4*4 = 16` is added.

The resulting list is `[4, 16]`.
</details>

---

### Problem 2.11

**Question:** What will be the output of the following code? Predict the output before running.

In [20]:
number = 12345
print(len(number))

TypeError: object of type 'int' has no len()

In [21]:
text_number = "12345"
print(len(text_number))

5


<details>
<summary>Hint</summary>
The `len()` function is designed to work with sequences (like strings, lists, tuples, dictionaries) to return their number of items, not with individual numeric types.
</details>

<details>
<summary>Solution</summary>

```
TypeError: object of type 'int' has no len()
5
```

</details>

<details>
<summary>Explanation</summary>
The `len()` function is used to get the length (number of items) of an object. It works for objects that are *sequences* or *collections* such as strings, lists, tuples, dictionaries, and sets. Integers are not sequences, so attempting to call `len()` on an integer directly results in a `TypeError`.

To get the number of digits in an integer, you would first need to convert it to a string, as shown in the second part of the example (`len(text_number)`).
</details>

---

## Functions, Tuples, Dictionaries, Exceptions, and Data Processing

This module delves into more advanced data structures, function behaviors, and error handling.

### Problem 3.1

**Question:** What will be the output of the following code? Predict the output before running.

In [22]:
my_tuple = (1, 2, 3)
my_tuple[0] = 99
print(my_tuple)

TypeError: 'tuple' object does not support item assignment

In [23]:
my_list = [1, 2, 3]
my_list[0] = 99
print(my_list)

[99, 2, 3]


<details>
<summary>Hint</summary>
Python data types are categorized as either mutable (changeable) or immutable (unchangeable) after creation. Tuples are immutable.
</details>

<details>
<summary>Solution</summary>

```
TypeError: 'tuple' object does not support item assignment
[99, 2, 3]
```

</details>

<details>
<summary>Explanation</summary>
This problem highlights the crucial difference between mutable and immutable data types in Python.
* **Tuples** are *immutable*. Once a tuple is created, you cannot change its elements, add new elements, or remove existing elements. Attempting to do so, like `my_tuple[0] = 99`, results in a `TypeError`.
* **Lists** are *mutable*. You can modify their elements, add, or remove items after creation, as demonstrated by `my_list[0] = 99` which successfully changes the first element.

Understanding mutability is fundamental for avoiding unexpected side effects in your code.
</details>

---

### Problem 3.2

**Question:** What will be the output of the following code? Predict the output before running.

In [24]:
list = [1, 2, 3]
print(list("hello"))
print(f"Caught an error: {e}")
del list

TypeError: 'list' object is not callable

<details>
<summary>Hint</summary>
Assigning a variable the same name as a built-in function or type will 'shadow' (hide) the original built-in.
</details>

<details>
<summary>Solution</summary>

```
TypeError: 'list' object is not callable
```

</details>

<details>
<summary>Explanation</summary>
Python has many built-in functions and types (like `list`, `str`, `int`, `print`, `input`). When you create a variable with the same name as a built-in, you are 'shadowing' that built-in. This means that within the scope where your variable is defined, the original built-in is no longer accessible by that name.

In this case, `list = [1, 2, 3]` reassigns the name `list` from the built-in list type to a new list object. Consequently, when `list("hello")` is called, Python tries to call the *variable* `list` (which is `[1, 2, 3]`) as if it were a function, leading to a `TypeError` because a list object is not callable.

It's best practice to avoid using built-in names for your variables to prevent such conflicts.
</details>

---

### Problem 3.3

**Question:** What will be the output of the following code? Predict the output before running.

In [25]:
value = None
if value == 0:
    print("Value is zero")
elif value == False:
    print("Value is False")
elif value is None:
    print("Value is None")
else:
    print("Value is something else")

Value is None


<details>
<summary>Hint</summary>
`None` is a unique data type in Python, representing the absence of a value. While it evaluates to `False` in a boolean context, it is not equal to `0` or `False`.
</details>

<details>
<summary>Solution</summary>

```
Value is None
```

</details>

<details>
<summary>Explanation</summary>
`None` is a special constant in Python that represents the absence of a value or a null value. It is its own distinct type (`NoneType`).

* `value == 0` is `False` because `None` is not numerically equal to `0`.
* `value == False` is `False` because `None` is not equal to the boolean `False`, even though `bool(None)` would return `False` (meaning `None` is 'falsy' in a boolean context).
* `value is None` is `True`. The `is` operator checks for object identity, and there is only one `None` object in Python, so `value` indeed refers to that specific `None` object.

Always use `is None` or `is not None` when checking for `None` values, rather than `==`.
</details>

---

### Problem 3.4

**Question:** What will be the output of the following code? Predict the output before running.

In [27]:
def greet_and_return_none(name):
    print(f"Hello, {name}!")

result = greet_and_return_none("Bob")
print(result)

Hello, Bob!
None


<details>
<summary>Hint</summary>
If a Python function does not explicitly use a `return` statement with a value, what does it implicitly return?
</details>

<details>
<summary>Solution</summary>

```
Hello, Bob!
Function returned: None
```

</details>

<details>
<summary>Explanation</summary>
A common misconception is that if a function prints something, it also returns that value. In Python, if a function reaches its end without an explicit `return` statement, it implicitly returns `None`. The `print()` function itself displays output to the console but returns `None`.

In this example, `greet_and_return_none` prints "Hello, Bob!" but does not have a `return` statement, so `result` is assigned `None`.

To return a value, you must use `return value`:
```python
def get_greeting(name):
    return f"Hello, {name}!"

message = get_greeting("Alice")
print(message) # Output: Hello, Alice!
```
</details>

---

### Problem 3.5

**Question:** What will be the type of `my_tuple` in each case? Predict the output before running.

In [28]:
my_tuple_1 = (10)
my_tuple_2 = (10,)
print(type(my_tuple_1))
print(type(my_tuple_2))

<class 'int'>
<class 'tuple'>


<details>
<summary>Hint</summary>
For Python to recognize a single-element tuple, it requires a trailing comma.
</details>

<details>
<summary>Solution</summary>

```
<class 'int'>
<class 'tuple'>
```

</details>

<details>
<summary>Explanation</summary>
When you define `my_tuple_1 = (10)`, the parentheses are interpreted as a grouping mechanism for the integer `10`, not as a tuple constructor. Therefore, `my_tuple_1` is simply an integer.

To explicitly define a single-element tuple, you *must* include a trailing comma, as in `my_tuple_2 = (10,)`. This comma tells Python that you intend to create a tuple containing one element.
</details>

---

### Problem 3.6

**Question:** What can you say about the order of elements when iterating over `my_dict` in Python 3.6+ versus older versions? Predict the output before running.

In [29]:
my_dict = {"apple": 1, "banana": 2, "cherry": 3}
for key in my_dict:
    print(key)

apple
banana
cherry


<details>
<summary>Hint</summary>
Python's dictionary implementation has evolved regarding insertion order.
</details>

<details>
<summary>Solution</summary>

* **Python 3.7+:** The output order is guaranteed to be `apple`, `banana`, `cherry`.
* **Python 3.6:** The output order is insertion-ordered (an implementation detail, not officially guaranteed until 3.7).
* **Python 3.5 and older:** The output order is *not* guaranteed and could be arbitrary.

</details>

<details>
<summary>Explanation</summary>
This is a common "gotcha" related to Python versioning.

* **Before Python 3.7:** Dictionaries were *unordered* collections. While CPython 3.6 did preserve insertion order as an implementation detail, it was not part of the language specification, meaning other Python implementations (like PyPy or Jython) were not required to maintain this order.

* **Python 3.7 and later:** The insertion-order preservation of dictionary elements became an official part of the Python language specification. This means that when you iterate over a dictionary, the keys (and values, and items) will be returned in the order they were inserted.

It's crucial for exam takers to know this distinction, especially if the exam environment might use an older Python version or if the question implies behavior across versions.
</details>

---

### Problem 3.7

**Question:** Will the following code run without error? If not, why? Predict the output before running.

In [30]:
def func(a, b=10, c):
    print(a, b, c)

func(1, 2, 3)

SyntaxError: parameter without a default follows parameter with a default (1719690628.py, line 1)

<details>
<summary>Hint</summary>
Consider the rule about non-default arguments following default arguments in function definitions.
</details>

<details>
<summary>Solution</summary>

```
SyntaxError: non-default argument follows default argument
```

</details>

<details>
<summary>Explanation</summary>
In Python, once you define a parameter with a default value (like `b=10`), all subsequent parameters in the function definition *must also* have default values. This rule exists to prevent ambiguity when calling the function. If `c` didn't have a default, how would Python know if an argument passed after `a` was meant for `b` or `c` if `b` could be skipped?

To fix this, `c` would also need a default value, or `b` would need to be moved after `c` if `c` is always required without a default.
Corrected example: `def func(a, c, b=10):` or `def func(a, b=10, c=20):`
</details>

---

### Problem 3.8

**Question:** What will be the output of the following code? Predict the output before running.

In [31]:
x = 10

def my_function():
    x = 5
    print(x)

my_function()
print(x)

5
10


<details>
<summary>Hint</summary>
Python's scope rules dictate that assigning to a variable inside a function creates a new local variable unless `global` is explicitly used.
</details>

<details>
<summary>Solution</summary>

```
5
10
```

</details>

<details>
<summary>Explanation</summary>
When `my_function()` is called, the line `x = 5` creates a *new local variable* named `x` within the scope of `my_function`. This local `x` shadows (hides) the global `x`. So, `print(x)` inside the function prints the local `x` (which is 5).

After the function finishes, the local `x` is destroyed. The `print(x)` outside the function refers to the global `x`, which was never modified and still holds its initial value of 10. If you wanted to modify the global `x` inside the function, you would need to use the `global` keyword:

```python
x = 10

def my_function():
    global x
    x = 5
    print(x)

my_function()
print(x)
```
</details>

---

### Problem 3.9

**Question:** What will be the output of the following code? Predict the output before running.

In [33]:
def add_item(item, my_list=[]):
    my_list.append(item)
    return my_list

print(add_item(1))
print(add_item(2))
print(add_item(3, [4, 5]))
print(add_item(6))

[1]
[1, 2]
[4, 5, 3]
[1, 2, 6]


<details>
<summary>Hint</summary>
Default arguments are evaluated *only once* when the function is defined, not every time the function is called. This can lead to unexpected behavior with mutable default arguments.
</details>

<details>
<summary>Solution</summary>

```
[1]
[1, 2]
[4, 5, 3]
[1, 2, 6]
```

</details>

<details>
<summary>Explanation</summary>
This is a classic Python "gotcha." When `my_list=[]` is defined as a default argument, the empty list `[]` is created *once* when the `add_item` function is first defined (i.e., when the Python interpreter reads the `def` statement).

* **`print(add_item(1))`**: `my_list` is the default `[]`. `1` is appended. Returns `[1]`.
* **`print(add_item(2))`**: `my_list` is *still* the same list object that was created initially. It now contains `[1]`. `2` is appended to *that same list*. Returns `[1, 2]`.
* **`print(add_item(3, [4, 5]))`**: Here, a *new* list `[4, 5]` is explicitly passed as `my_list`. `3` is appended to *this new list*. Returns `[4, 5, 3]`. The original default list is unaffected.
* **`print(add_item(6))`**: Again, `my_list` defaults to the *original* list object, which now contains `[1, 2]`. `6` is appended to it. Returns `[1, 2, 6]`.

To avoid this, a common pattern is to use `None` as the default and initialize the list inside the function:

```python
def add_item_corrected(item, my_list=None):
    if my_list is None:
        my_list = []
    my_list.append(item)
    return my_list

print(add_item_corrected(1))
print(add_item_corrected(2))
print(add_item_corrected(3, [4, 5]))
print(add_item_corrected(6))
```
</details>

---

### Problem 3.10

**Question:** What will be the output of the following code? Predict the output before running.

In [35]:
try:
    print("Inside try block - Line 1")
    x = 1 / 0
    print("Inside try block - Line 2")
except ZeroDivisionError:
    print("Caught ZeroDivisionError")
except:
    print("Caught other error")
finally:
    print("Finally block executed")

Inside try block - Line 1
Caught ZeroDivisionError
Finally block executed


<details>
<summary>Hint</summary>
Code inside the `try` block executes line by line until an error occurs. Once an error occurs, the rest of the `try` block is skipped, and control jumps to the appropriate `except` block. The `finally` block always executes.
</details>

<details>
<summary>Solution</summary>

```
Inside try block - Line 1
Caught ZeroDivisionError
Finally block executed
```

</details>

<details>
<summary>Explanation</summary>
1.  `print("Inside try block - Line 1")` executes successfully.
2.  `x = 1 / 0` raises a `ZeroDivisionError`. At this point, the execution of the `try` block immediately stops.
3.  Python looks for an `except` block that can handle `ZeroDivisionError`. It finds `except ZeroDivisionError:`.
4.  The code inside that `except` block, `print("Caught ZeroDivisionError")`, executes.
5.  Regardless of whether an exception occurred or was handled, the `finally` block always executes. So, `print("Finally block executed")` runs.
</details>

---

### Problem 3.11

**Question:** Will the following code run without error? If not, why? Predict the output before running.

In [36]:
in = 10
print(in)

SyntaxError: invalid syntax (2172383762.py, line 1)

<details>
<summary>Hint</summary>
Python has a set of reserved words that have special meanings and cannot be used as identifiers (variable names, function names, etc.).
</details>

<details>
<summary>Solution</summary>

```
SyntaxError: invalid syntax
```

</details>

<details>
<summary>Explanation</summary>
`in` is a reserved keyword in Python, used to check if a value is present in a sequence. You cannot use reserved keywords as variable names, function names, or any other identifiers. Doing so results in a `SyntaxError`, which is caught by the Python parser *before* the code even begins to execute.
</details>

---

### Problem 3.12

**Question:** What will be the output of the following code? This type of error occurs during the parsing phase and typically prevents the script from running at all. Predict the output before running.

In [39]:
try:
    print('Hello')
    print('Hello')
except SyntaxError:
    print("Caught SyntaxError")

Hello
Hello


<details>
<summary>Hint</summary>
A `SyntaxError` in the main structure of your code is detected before runtime, meaning `try-except` blocks cannot catch it.
</details>

<details>
<summary>Solution</summary>

```
SyntaxError: invalid syntax (or similar, depending on Python version and exact error)
```

(This error occurs *before* the `try-except` block can even be entered, preventing execution.)

</details>

<details>
<summary>Explanation</summary>
A `SyntaxError` occurs when the Python parser detects an invalid syntax in the code's structure itself. This type of error is found during the initial parsing phase, *before* the program actually begins to execute. Because the error prevents the code from even being interpreted, the `try-except` block (which is a runtime construct) is never reached. The program will simply terminate with a `SyntaxError`.

</details>

---

### Problem 3.13

**Question:** What will be the output of the following code? This type of `SyntaxError` occurs during runtime and can be caught. Predict the output before running.

In [40]:
try:
    eval("print('Hello')")
    eval("print('Hello'")
except SyntaxError:
    print("Caught SyntaxError from eval")
except Exception as e:
    print(f"Caught other error: {e}")

Hello
Caught SyntaxError from eval


<details>
<summary>Hint</summary>
When functions like `eval()` or `exec()` encounter a syntax error within the string they are processing, that error is raised as a runtime exception.
</details>

<details>
<summary>Solution</summary>

```
Hello
Caught SyntaxError from eval
```

</details>

<details>
<summary>Explanation</summary>
When functions like `eval()` or `exec()` attempt to execute a string that *contains* a syntax error, that `SyntaxError` is raised *at runtime* by the function itself. In these specific scenarios, a `try-except` block *can* catch it.

In this example:
1.  `eval("print('Hello')")` executes successfully, printing "Hello".
2.  `eval("print('Hello'")` attempts to execute a string with a syntax error (missing closing parenthesis for `print`).
3.  The `eval()` function raises a `SyntaxError` at this point.
4.  This `SyntaxError` is then caught by our `except SyntaxError:` block, and "Caught SyntaxError from eval" is printed.

The key distinction is whether the syntax error prevents the *program from starting* (uncatchable, as seen in Problem 3.12) or occurs *during execution* of a function like `eval()` (catchable).
</details>

---

### Problem 3.14

**Question:** What will be the output of the following code? Predict the output before running.

In [41]:
list1 = [1, 2, 3]
list2 = [1, 2, 3]
list3 = list1

print(list1 == list2)
print(list1 is list2)
print(list1 == list3)
print(list1 is list3)

a = 1
b = 1
print(a is b)

x = 300
y = 300
print(x is y)

True
False
True
True
True
False


<details>
<summary>Hint</summary>
`==` checks for *equality* (do they have the same value?). `is` checks for *identity* (do they refer to the exact same object in memory?). Python optimizes small integers.
</details>

<details>
<summary>Solution</summary>

```
True
False
True
True
True
False
```

</details>

<details>
<summary>Explanation</summary>

* `list1 == list2`: `True`. Both lists contain the same elements in the same order, so their *values* are equal.
  
* `list1 is list2`: `False`. `list1` and `list2` are two separate list objects in memory, even though they have identical content. The `is` operator checks if they are the *exact same object*.
  
* `list1 == list3`: `True`. `list3` is an alias for `list1`, so they refer to the same object, and thus their values are equal.
  
* `list1 is list3`: `True`. `list3` was assigned `list1`, meaning both variables point to the exact same list object in memory.

* `a = 1`, `b = 1`, `print(a is b)`: `True`. This is due to a CPython optimization where small integers (typically from -5 to 256) are 'interned' or 'cached'. This means that when you create multiple variables with these small integer values, Python often points them all to the *same* immutable object in memory for efficiency.
  
* `x = 300`, `y = 300`, `print(x is y)`: `False`. For larger integers, Python typically creates distinct objects even if their values are the same. Therefore, `x` and `y` refer to different objects in memory.

Always use `==` for value comparison and `is` for identity comparison.
</details>

---

## Conclusion

We hope this guide has shed light on some of the trickier aspects of Python that are frequently tested in the PCEP exam. By understanding these common errors and their underlying concepts, you'll be better prepared to tackle similar problems on your exam. Keep practicing, and good luck!